### ISSUE TITLE
LDM 訓練 Validation Loss 異常飆升與字形結構崩壞問題請益

### ISSUE THREAD

**Issue by XLOIEUD** [wangwenho/HanziGen#8]
您好，感謝您開源這個優秀的 HanziGen 專案！
我在基於貴專案的架構嘗試訓練自集的中文字體（集合自古代法帖，書法風格，字重不定且筆劃處理與規範漢字有區別）的中文字體時，遇到了一些 LDM 訓練上的瓶頸，希望能向您請教可能的除錯方向。

1. 當前狀況與問題
訓練環境與參數
- Base Model: 依照 Repo 提供的預設配置
- 硬體資源: RTX 3060laptop, 6G VRAM)
- Batch Size: 16
- 參與訓練的字符數：1809/1900

現象：訓練至 200 Epochs 左右，觀察 Sample 結果，發現生成的圖像在筆劃風格上與 Target 接近，LPIPS 也有下降趨勢，但在字形結構上基本沒法使用。
問題：我正在訓練 Latent Diffusion Model (LDM) 部分。目前的現象是：模型極易過擬合 ，且 漢字的結構發生崩壞。幾乎所有測試集中和推理的的文字結構都不正確（除極少部分筆劃較少，結構簡單的漢字，且另外大多數筆劃簡單的漢字也出現結構錯誤）。反饋于參數則呈現出如下特點：

- Loss/val 曲線異常：Validation Loss (MSE) 呈現明顯的「U 型」反彈。大約在 100 Epoch 左右達到最低點，隨後飆升，似乎出現了嚴重的過擬合。
- LPIPS 曲線異常： Loss 的劇烈波動相比，LPIPS 的下降伴隨震蕩地緩慢下降，且在 Loss 反彈後也無法降至理想水準（最低點維持在0.2左右）。
- 生成結果崩壞：分別取用楷體為 Reference 和 Jigmo 為 Reference 的LPIPS最低點的模型進行推論，生成的字符結構是嚴重錯誤的（缺筆畫、部件錯位），看起來模型似乎無法正確利用 Reference 圖片的骨架。

2. 已排除的因素

為了定位問題，我已經進行了檢查，初步排除了 VQ-VAE 與資料對齊的問題。
VQ-VAE 運作正常：我檢查過 VQ-VAE 的重建影像，對於目標字體的還原度非常高，筆畫與細節都很清晰，因此 Latent Space 的壓縮與解壓縮應該是沒問題的。

3. 嘗試過的解決方案

針對上述過擬合與結構崩壞的問題，我嘗試了以下調整，但效果有限：

1. 嘗試新方案：更換為風格相近的行楷作為 Reference。我挑選了一款骨架與 Target 更為接近、且筆畫粗細相當的楷體。
- 資料量： 由於取交集，資料量從 1900 降至 1809 字。
- 結果： Loss 曲線依然震盪明顯，結構依然錯誤，甚至錯誤更加嚴重。
2. 降低 Learning Rate：將 LR 從 5e-4 下修至 1e-4 。結構學習效果仍不佳。

目前我可以想到的調整方案：
1. 調整 Batch Size：我全程訓練使用默認 Batch Size = 16 ，不知道調高Batch Size 是否有所幫助。
2. 優化器調整：嘗試使用 AdamW 並加入 Weight Decay。

4. 附圖與日誌

- Loss / LPIPS 曲線圖：

![Image](https://github.com/user-attachments/assets/ba982afa-18e4-41fb-87a1-07560dc4b7d1)

（使用風格接近的楷體作爲Reference）

![Image](https://github.com/user-attachments/assets/0299e5e5-5fe4-4dfa-bb54-3d1ad6f434c6)

（使用Jigmo作爲Reference）

- 失敗的生成樣本：

![Image](https://github.com/user-attachments/assets/1ae164da-825f-408b-95ff-da90a9a196a8)

（使用風格接近的楷體作爲Reference）

![Image](https://github.com/user-attachments/assets/ce06b10f-4dfe-4554-a704-2ef74527709a)

（使用Jigmo作爲Reference）

- VQ-VAE 重建樣本 (VQ-VAE Reconstructions)：

![Image](https://github.com/user-attachments/assets/6900bd87-a6e6-408a-bf46-ff71b8fd29a6)

（使用風格接近的楷體作爲Reference）

![Image](https://github.com/user-attachments/assets/24a66ad8-c080-4579-b9bf-e79197bc35fa)

（使用Jigmo作爲Reference）

5. 請教

想請問作者或各位先進：

1. 在訓練行楷這類筆劃特點複雜的字體時，LDM 是否需要特殊的參數調整？
2. Reference的選擇是否與Target的風格愈接近愈好？還是説以宋體或黑體為宜？
3. 這種 Loss 飛漲但 LPIPS 尚未收斂的現象，是否暗示模型的某個機制可能沒有正確運作？有沒有建議的 Debug 方式？

非常感謝您的撥冗閱讀與指導！

**Comment by wangwenho** [comment]
您好，非常感謝您使用本專案，並提供了如此詳盡的實驗數據、Log 與生成樣本。針對您提出的三個問題，以下是我的初步分析與建議：

1. 在訓練行楷這類筆劃特點複雜的字體時，LDM 是否需要特殊的參數調整？

   1. 這屬於本專案在架構設計上的先天限制。由於本專案的模型參數容量較小，且針對字形結構控制的機制設計較為簡潔；加上訓練資料集較小，模型在學習字體風格特徵時，較難兼顧字體結構的轉換，導致泛化能力較差。

      此一短板在面對手寫、行草、行楷等「筆劃連筆多、結構變異性大」的字體時，會被顯著放大。在現有的架構下，單純透過調整超參數可能難以從根本解決結構崩壞的問題。

      至於您提到的調整批次大小或更換優化器種類，雖然是常見的優化手段，但在模型容量與結構控制機制受限的前提下，這些調整對於解決結構崩壞的幫助可能相當有限。

2. Reference 的選擇是否與 Target 的風格愈接近愈好？還是説以宋體或黑體為宜？

   1. 關於參考字體的選擇，根據我先前的實驗經驗，使用標準的黑體或明體（宋體）通常優於風格相近的字體。

      參考字體與目標字體風格相近雖然有助於學習，但在訓練資料集不足的情況下，使用具備穩定筆畫與標準結構規則的字體（如黑體、明體），能大幅減少模型在推斷字形骨架時的歧義，為模型提供更明確的結構指引。反之，若參考字體本身帶有強烈風格，反而可能干擾模型對「骨架」與「風格」的特徵解離。

3. 這種 Loss 飛漲但 LPIPS 尚未收斂的現象，是否暗示模型的某個機制可能沒有正確運作？有沒有建議的 Debug 方式？

   1. 請放心，這是圖像生成模型訓練中十分常見的現象。在本專案中，LDM 的訓練目標是最小化雜訊預測的 MSE，這代表模型對機率分佈的擬合程度；而我們儲存權重的依據是驗證集的 LPIPS，這代表人類視覺上的感知品質。

      在訓練中後期，經常會觀察到驗證 MSE 曲線呈現 U 型反彈（這在統計上代表著過擬合，即模型開始死記硬背訓練集的雜訊特徵，導致對驗證集的預測誤差變大）。然而，擴散模型的生成過程包含多步去噪，即便單步的雜訊預測誤差略微上升，模型在「捕捉高頻細節」和「紋理生成」的能力上可能仍在進步，因此反映在 LPIPS 上會看到數值持續下降，生成的圖像在視覺上也更自然。

      在生成任務中，我們應優先關注「最終生成圖像的感知品質」，而非僅止於「去噪過程的數值精確度」。建議您主要依據 LPIPS 曲線的趨勢來判斷收斂狀況。當 LPIPS 數值進入平原期後，請直接結合肉眼觀察來檢視驗證集的生成圖像；若發現圖像結構尚未崩壞，且生成效果已達預期，即可視為訓練完成並停止訓練。

希望以上回覆對您有所幫助 :)
